ssl_calls: support OpenSSL 1.1.0 API for setting DH params

speidy opened the pull request:
OpenSSL 1.1.0 makes many structs opaque now, so we can't set its values directly anymore.

Also:
- fix memory leak introduced by PR #1024 in `get_dh2236()`.
- move `SSL_CTX` object creation check to its right place. use `log_message` for logging.

metalefty commented on common/ssl_calls.c:
```diff
@@ -641,25 +689,42 @@ ssl_tls_accept(struct ssl_tls *self, long ssl_protocols,
     if (g_strlen(tls_ciphers) > 1)
     {
-        log_message(LOG_LEVEL_TRACE, "ssl_tls_accept: tls_ciphers=%s", tls_ciphers);
+        log_message(LOG_LEVEL_TRACE, "ssl_tls_accept: tls_ciphers=%s", \
```
backslash is not needed here.

speidy replied:
k

speidy commented:
Meta, can you test on OpenSSL < 1.0.x (1.0.2)?

Thx
בתאריך יום א׳, 18 במרץ 2018 ב-16:12 מאת metalefty <notifications@github.com
>:

> *@metalefty* commented on this pull request.
> ------------------------------
>
> In common/ssl_calls.c
> <https://github.com/neutrinolabs/xrdp/pull/1063#discussion_r175290570>:
>
> >  #if defined(SSL_CTX_set_ecdh_auto)
>      SSL_CTX_set_ecdh_auto(self->ctx, 1);
>  #endif
>
>      if (g_strlen(tls_ciphers) > 1)
>      {
> -        log_message(LOG_LEVEL_TRACE, "ssl_tls_accept: tls_ciphers=%s", tls_ciphers);
> +        log_message(LOG_LEVEL_TRACE, "ssl_tls_accept: tls_ciphers=%s", \
>
> backslash is not needed here.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neutrinolabs/xrdp/pull/1063#pullrequestreview-104798633>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADTH1H_e_RHtXSjN5cG1zbypLesw6hRtks5tfms5gaJpZM4Sutmj>
> .
>
-- 
Idan Freiberg

PGP FP: 8108 7EC9 806E 4980 75F2  72B3 8AD3 2D04 337B 1F18

speidy commented:
maybe we need to add travis build with multi ssl versions

metalefty commented:
Yes. We need to switch Docker-based travis build. I already do it in RH-Matic to test multiple CentOS version. The bare travis build environment is Ubuntu 14.04. Can't test vsock, too.

metalefty commented:
Can you resolve conflict before I test with OpenSSL 1.0.x?

speidy commented:
@metalefty docker builds will be great. I'll work on that too.

About the conflict, I was thinking maybe the static DH groups are actually ok :)
But I want to have some coding style changes

speidy commented:
i'll add them to this PR

speidy commented:
ok, added DH groups back and changed the function name.
can you review again? if its fine we can merge.

Iv'e tested on Ubuntu 16.04 (openssl 1.0.2) and archlinux (openssl 1.1.0g)

metalefty commented on common/ssl_calls.c:
```diff
@@ -643,27 +690,42 @@ ssl_tls_accept(struct ssl_tls *self, long ssl_protocols,
+
+    /* set DH parameters */
+    DH *dh = ssl_get_dh2236();
+    if (NULL == dh)
```
Yoda notation is sometimes helpful but using it only here is inconsistent.

metalefty commented:
Overall, logic looks good. Let me test it later.

metalefty commented on common/ssl_calls.c:
```diff
@@ -643,27 +690,42 @@ ssl_tls_accept(struct ssl_tls *self, long ssl_protocols,
     {
-        g_writeln("SSL_CTX_set_tmp_dh failed");
+        log_message(LOG_LEVEL_ERROR, "ssl_tls_accept: SSL_CTX_set_tmp_dh \
+        	failed");
```
tab. what about like this?

```
log_message(LOG_LEVEL_ERROR,
            "ssl_tls_accept: SSL_CTX_set_tmp_dh failed");
```

metalefty commented on common/ssl_calls.c:
```diff
@@ -673,12 +735,6 @@ ssl_tls_accept(struct ssl_tls *self, long ssl_protocols,
-        g_writeln("ssl_tls_accept: SSL_CTX_new failed");
-        return 1;
-    }
-
```
Good catch 👍 This check should be immedeately after new.

metalefty commented:
Tested, working fine. If no more concerns, let's ship it to v0.9.6.

speidy commented:
Lets merge
בתאריך יום ה׳, 22 במרץ 2018 ב-4:59 מאת metalefty <notifications@github.com>:

> Tested, working fine. If no more concerns, let's ship it to v0.9.6.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neutrinolabs/xrdp/pull/1063#issuecomment-375163279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADTH1KL_lXIztIxUNLY8cHwQ86X2piW3ks5tgxNtgaJpZM4Sutmj>
> .
>
-- 
Idan Freiberg

PGP FP: 8108 7EC9 806E 4980 75F2  72B3 8AD3 2D04 337B 1F18
